### PR TITLE
Accept `Enum::<Param>::Variant {}` over `Enum::Variant::<Param> {}`

### LINKED ISSUE #69356: Possibly incorrect syntax `Enum::<Type, Params>::Variant` allowed since 1.33
A coworker recently identified the following case:

```
pub enum Foo<'a, T> {
    Struct {},
    Tuple(),
    Unit,
    Usage(&'a T),
}
pub fn main() {
    let foo = Foo::<String>::Unit;
    let foo = Foo::<String>::Tuple();
    let foo = Foo::<String>::Struct {};
}
```

[This used to fail because of the lack of implicit bound `T: 'a`](https://godbolt.org/z/-zUiCa):
```
error[E0309]: the parameter type `T` may not live long enough
 --> <source>:5:11
  |
1 | pub enum Foo<'a, T> {
  |                  - help: consider adding an explicit lifetime bound `T: 'a`...
...
5 |     Usage(&'a T),
  |           ^^^^^
  |
note: ...so that the reference type `&'a T` does not outlive the data it points at
 --> <source>:5:11
  |
5 |     Usage(&'a T),
  |           ^^^^^
```

[After fixing that, it complains about the type params](https://godbolt.org/z/HwCIEP):

```
error[E0109]: type parameters are not allowed on this type
 --> <source>:8:21
  |
8 |     let foo = Foo::<String>::Unit;
  |                     ^^^^^^ type parameter not allowed

error[E0109]: type parameters are not allowed on this type
 --> <source>:9:21
  |
9 |     let foo = Foo::<String>::Tuple();
  |                     ^^^^^^ type parameter not allowed

error[E0109]: type parameters are not allowed on this type
  --> <source>:10:21
   |
10 |     let foo = Foo::<String>::Struct {};
   |                     ^^^^^^ type parameter not allowed
```

[Starting in 1.33, all but the `Struct` path are accepted](https://godbolt.org/z/EO_SBq):

```
error[E0110]: lifetime arguments are not allowed on this entity
  --> <source>:10:15
   |
10 |     let foo = Foo::<String>::Struct {};
   |               ^^^^^^^^^^^^^^^^^^^^^ lifetime argument not allowed

error[E0107]: wrong number of lifetime arguments: expected 1, found 0
  --> <source>:10:15
   |
10 |     let foo = Foo::<String>::Struct {};
   |               ^^^^^^^^^^^^^^^^^^^^^ expected 1 lifetime argument
```

[The correct code for this would of course be](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=2afe55bc512078a34ba3e37409d74902):

```
pub enum Foo<'a, T> {
    Struct {},
    Tuple(),
    Unit,
    Usage(&'a T),
}
pub fn main() {
    let foo = Foo::Unit::<String>;
    let foo = Foo::Tuple::<String>();
    let foo = Foo::Struct::<String> {};
}
```

My questions:

- Was this change in behavior (accepting some of these paths) intended?
- If so, was it intended to deny only _one_ of these cases?
- If this change wasn't intended, should we try to reintroduce it or grandfather it in?

Depending on the answer to those questions, the solution to this ticket could be:

- Make the compiler accept the `let foo = Foo::Struct::<String> {};` case
- Deny all cases with a suggestion pointing at the right syntax

cc @Centril

### PR DESCRIPTION
**Pull request by estebank**
Fix https://github.com/rust-lang/rust/issues/69356.

r? @petrochenkov @eddyb @Centril

### CODE REVIEW

File: src/test/ui/nll/user-annotations/pattern_substs_on_brace_enum_variant.rs [hunk 1]
```diff
@@ -14,8 +14,38 @@ fn in_match() {
...
+    let y = 22;
+    let foo = Foo::Bar { field: &y };
+    let Foo::Bar { field: _z } = foo;
+}
+
+fn in_match_3() {
+    let y = 22;
+    let foo = Foo::Bar { field: &y };
+    match foo {
+        Foo::Bar { field: _z } => {}
```
Comment (estebank) [inline src/test/ui/nll/user-annotations/pattern_substs_on_brace_enum_variant.rs]: This should be passing and it isn't, I'm guessing I need to be using `Value` where I'm using `Type` somewhere?

Comment (Centril) [reply src/test/ui/nll/user-annotations/pattern_substs_on_brace_enum_variant.rs]: cc @matthewjasper

Comment (estebank) [reply src/test/ui/nll/user-annotations/pattern_substs_on_brace_enum_variant.rs]: CC @matthewjasper: @eddyb mentioned that the "user type annotation" system for MIR borrowck might be needing adjustment for these changes. Could you point me in the right direction?

Comment (estebank) [reply src/test/ui/nll/user-annotations/pattern_substs_on_brace_enum_variant.rs]:
```
[DEBUG rustc_mir::borrow_check::type_check] check_user_type_annotations: user_type_annotations=[CanonicalUserTypeAnnotation { user_ty: Canonical { max_universe: U0, variables: [], value: TypeOf(DefId(0:3 ~ pattern_substs_on_brace_enum_variant[317d]::Foo[0]), UserSubsts { substs: [ReStatic], user_self_ty: None }) }, span: src/test/ui/nll/user-annotations/pattern_substs_on_brace_enum_variant.rs:40:15: 40:37, inferred_ty: Foo<'_> }, CanonicalUserTypeAnnotation { user_ty: Canonical { max_universe: U0, variables: [], value: TypeOf(DefId(0:3 ~ pattern_substs_on_brace_enum_variant[317d]::Foo[0]), UserSubsts { substs: [ReStatic], user_self_ty: None }) }, span: src/test/ui/nll/user-annotations/pattern_substs_on_brace_enum_variant.rs:41:9: 41:31, inferred_ty: Foo<'_> }, CanonicalUserTypeAnnotation { user_ty: Canonical { max_universe: U0, variables: [], value: TypeOf(DefId(0:3 ~ pattern_substs_on_brace_enum_variant[317d]::Foo[0]), UserSubsts { substs: [ReStatic], user_self_ty: None }) }, span: src/test/ui/nll/user-annotations/pattern_substs_on_brace_enum_variant.rs:41:9: 41:31, inferred_ty: Foo<'_> }]
```

Comment (estebank) [reply src/test/ui/nll/user-annotations/pattern_substs_on_brace_enum_variant.rs]:
```
[DEBUG rustc_mir::borrow_check::renumber] visit_region: region='_#2r
[DEBUG rustc_mir::borrow_check::renumber] visit_region(region=ReErased, location=bb0[7])
[DEBUG rustc_mir::borrow_check::renumber] renumber_regions(value=ReErased)
[DEBUG rustc_mir::borrow_check::renumber] visit_region: region='_#3r
[DEBUG rustc_mir::borrow_check::renumber] visit_substs(substs=[ReStatic], location=bb0[8])
[DEBUG rustc_mir::borrow_check::renumber] renumber_regions(value=[ReStatic])
[DEBUG rustc_mir::borrow_check::renumber] visit_substs: substs=['_#4r]
[DEBUG rustc_mir::borrow_check::renumber] renumber_regions(value=&'static u32)
[DEBUG rustc_mir::borrow_check::renumber] visit_ty(ty=(), ty_context=ReturnTy(SourceInfo { span: src/test/ui/nll/user-annotations/pattern_substs_on_brace_enum_variant.rs:38:1: 42:2, scope: scope[0] }))
[DEBUG rustc_mir::borrow_check::renumber] renumber_regions(value=())
[DEBUG rustc_mir::borrow_check::renumber] visit_ty: ty=()
[DEBUG rustc_mir::borrow_check::renumber] visit_ty(ty=(), ty_context=LocalDecl { local: _0, source_info: SourceInfo { span: src/test/ui/nll/user-annotations/pattern_substs_on_brace_enum_variant.rs:38:15: 38:15, scope: scope[0] } })
[DEBUG rustc_mir::borrow_check::renumber] renumber_regions(value=())
[DEBUG rustc_mir::borrow_check::renumber] visit_ty: ty=()
[DEBUG rustc_mir::borrow_check::renumber] visit_ty(ty=u32, ty_context=LocalDecl { local: _1, source_info: SourceInfo { span: src/test/ui/nll/user-annotations/pattern_substs_on_brace_enum_variant.rs:39:9: 39:10, scope: scope[0] } })
[DEBUG rustc_mir::borrow_check::renumber] renumber_regions(value=u32)
[DEBUG rustc_mir::borrow_check::renumber] visit_ty: ty=u32
[DEBUG rustc_mir::borrow_check::renumber] visit_ty(ty=Foo<'static>, ty_context=LocalDecl { local: _2, source_info: SourceInfo { span: src/test/ui/nll/user-annotations/pattern_substs_on_brace_enum_variant.rs:40:9: 40:12, scope: scope[1] } })
[DEBUG rustc_mir::borrow_check::renumber] renumber_regions(value=Foo<'static>)
[DEBUG rustc_mir::borrow_check::renumber] visit_ty: ty=Foo<'_>
[DEBUG rustc_mir::borrow_check::renumber] visit_ty(ty=&'static u32, ty_context=LocalDecl { local: _3, source_info: SourceInfo { span: src/test/ui/nll/user-annotations/pattern_substs_on_brace_enum_variant.rs:40:33: 40:35, scope: scope[0] } })
[DEBUG rustc_mir::borrow_check::renumber] renumber_regions(value=&'static u32)
[DEBUG rustc_mir::borrow_check::renumber] visit_ty: ty=&u32
```

vs before the change

```
[DEBUG rustc_mir::borrow_check::renumber] visit_region: region='_#2r
[DEBUG rustc_mir::borrow_check::renumber] visit_region(region=ReErased, location=bb0[7])
[DEBUG rustc_mir::borrow_check::renumber] renumber_regions(value=ReErased)
[DEBUG rustc_mir::borrow_check::renumber] visit_region: region='_#3r
[DEBUG rustc_mir::borrow_check::renumber] visit_substs(substs=[ReScope(Remainder { block: 26, first_statement_index: 1})], location=bb0[8])
[DEBUG rustc_mir::borrow_check::renumber] renumber_regions(value=[ReScope(Remainder { block: 26, first_statement_index: 1})])
[DEBUG rustc_mir::borrow_check::renumber] visit_substs: substs=['_#4r]
[DEBUG rustc_mir::borrow_check::renumber] renumber_regions(value=&u32)
[DEBUG rustc_mir::borrow_check::renumber] visit_ty(ty=(), ty_context=ReturnTy(SourceInfo { span: src/test/ui/nll/user-annotations/pattern_substs_on_brace_enum_variant.rs:5:1: 10:2, scope: scope[0] }))
[DEBUG rustc_mir::borrow_check::renumber] renumber_regions(value=())
[DEBUG rustc_mir::borrow_check::renumber] visit_ty: ty=()
[DEBUG rustc_mir::borrow_check::renumber] visit_ty(ty=(), ty_context=LocalDecl { local: _0, source_info: SourceInfo { span: src/test/ui/nll/user-annotations/pattern_substs_on_brace_enum_variant.rs:5:13: 5:13, scope: scope[0] } })
[DEBUG rustc_mir::borrow_check::renumber] renumber_regions(value=())
[DEBUG rustc_mir::borrow_check::renumber] visit_ty: ty=()
[DEBUG rustc_mir::borrow_check::renumber] visit_ty(ty=u32, ty_context=LocalDecl { local: _1, source_info: SourceInfo { span: src/test/ui/nll/user-annotations/pattern_substs_on_brace_enum_variant.rs:6:9: 6:10, scope: scope[0] } })
[DEBUG rustc_mir::borrow_check::renumber] renumber_regions(value=u32)
[DEBUG rustc_mir::borrow_check::renumber] visit_ty: ty=u32
[DEBUG rustc_mir::borrow_check::renumber] visit_ty(ty=Foo<'_>, ty_context=LocalDecl { local: _2, source_info: SourceInfo { span: src/test/ui/nll/user-annotations/pattern_substs_on_brace_enum_variant.rs:7:9: 7:12, scope: scope[1] } })
[DEBUG rustc_mir::borrow_check::renumber] renumber_regions(value=Foo<'_>)
[DEBUG rustc_mir::borrow_check::renumber] visit_ty: ty=Foo<'_>
[DEBUG rustc_mir::borrow_check::renumber] visit_ty(ty=&u32, ty_context=LocalDecl { local: _3, source_info: SourceInfo { span: src/test/ui/nll/user-annotations/pattern_substs_on_brace_enum_variant.rs:7:33: 7:35, scope: scope[0] } })
[DEBUG rustc_mir::borrow_check::renumber] renumber_regions(value=&u32)
[DEBUG rustc_mir::borrow_check::renumber] visit_ty: ty=&u32
[DEBUG rustc_mir::borrow_check::renumber] visit_ty(ty=&u32, ty_context=LocalDecl { local: _4, source_info: SourceInfo { span: src/test/ui/nll/user-annotations/pattern_substs_on_brace_enum_variant.rs:7:33: 7:35, scope: scope[0] } })
[DEBUG rustc_mir::borrow_check::renumber] renumber_regions(value=&u32)
```

**Comment by petrochenkov** [comment]
So, I remember the rule here - paths in "value" contexts get all their omitted generic arguments filled with inference variables (`_`), but paths in "type" contexts get all their omitted generic arguments filled with defaults or errors if some necessary defaults don't exist.

All the examples in https://github.com/rust-lang/rust/issues/69356 and in the tests are in "value" context.

I did touch the implementation of this at some point, but it was rewritten so many times since then that I have no idea how this works now.

**Review by eddyb — APPROVED** [review]

File: src/librustc_ast_lowering/path.rs [hunk 2]
```diff
@@ -55,9 +55,6 @@ impl<'a, 'hir> LoweringContext<'a, 'hir> {
                         Res::Def(DefKind::AssocTy, def_id) if i + 2 == proj_start => {
                             Some(parent_def_id(self, def_id))
                         }
-                        Res::Def(DefKind::Variant, def_id) if i + 1 == proj_start => {
-                            Some(parent_def_id(self, def_id))
-                        }
```
Comment (petrochenkov) [inline src/librustc_ast_lowering/path.rs]: Why is this deleted?
This way we don't get the correct number of lifetime parameters in the `Enum` segment in `some::prefix::Enum::Variant`.

File: src/librustc_typeck/astconv.rs [hunk 3]
```diff
@@ -2535,7 +2549,7 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
             | Res::Def(DefKind::ForeignTy, did) => {
                 assert_eq!(opt_self_ty, None);
                 self.prohibit_generics(path.segments.split_last().unwrap().1);
-                self.ast_path_to_ty(span, did, path.segments.last().unwrap())
+                self.ast_path_to_ty(span, did, path.segments.last().unwrap(), generic_arg_position)
             }
             Res::Def(kind @ DefKind::Variant, def_id) if permit_variants => {
```
Comment (petrochenkov) [inline src/librustc_typeck/astconv.rs]:
```suggestion
            Res::Def(kind @ DefKind::Variant, def_id) if generic_arg_position == GenericArgPosition::Value => {
```
(Used only once.)

**Comment by petrochenkov** [comment]
Is @varkor the author of the current implementation?

File: src/librustc_ast_lowering/path.rs (hunk 2 above)
Comment (eddyb) [reply src/librustc_ast_lowering/path.rs]: The lifetimes are for elision, so the full path would need to be a type (e.g. in a function signature) for this to matter.

If we want to be future-compat with variant types, we can keep this and make it depend on which segment has generic args, like typeck does.

Comment (estebank) [reply src/librustc_ast_lowering/path.rs]: Changed it in a backwards compatible way. Now the default behavior will remain the same _unless_ the `enum` path segment has arguments, and in that case _only that one_ will have the `Some(parent_def_id)`. This keeps the current behavior while also allowing the newer syntax to work correctly.

**Comment by estebank** [comment]
I've got a commit to make the lint allow by default if that is preferable.

File: src/librustc_ast_lowering/path.rs [hunk 4]
```diff
@@ -55,7 +56,21 @@ impl<'a, 'hir> LoweringContext<'a, 'hir> {
...
+                            type_param_in_enum = true;
+                            Some(parent_def_id(self, def_id))
+                        }
+                        Res::Def(DefKind::Variant, def_id)
+                            if i + 1 == proj_start
+                                && (segment.args.is_some() || !type_param_in_enum) =>
+                        {
+                            // Handle `Enum::Variant::<Param> {}`.
                             Some(parent_def_id(self, def_id))
                         }
```
Comment (eddyb) [inline src/librustc_ast_lowering/path.rs]: Two things here:
1. you can avoid mutable state by looking at `p.segments`
  * btw I doubt a bit that `.args.is_some()` is the correct way to check this, it might be true even for `::<>` which I'm not sure if it's compatible with the similar check in typeck
2. these only make sense if the param mode is explicit, which it can never be for `Expr::Struct` right now (`<Enum>::Variant {...}` is not valid syntax, and `Enum::Variant` is not valid in type paths *yet*, only one of these two could actually be relevant here)
  * the comments referring to `Enum::Variant {...}` are therefore inaccurate IMO
  * this needs some extra checks, probably for the param mode being explicit
  * it's really mostly forward compat for `enum` variant types so you could just as well add an assert that `DefKind::Variant` parts are never `ParamMode::Explicit`

File: src/librustc_session/lint/builtin.rs [hunk 5]
```diff
@@ -531,6 +538,7 @@ declare_lint_pass! {
         PATTERNS_IN_FNS_WITHOUT_BODY,
         MISSING_FRAGMENT_SPECIFIER,
         LATE_BOUND_LIFETIME_ARGUMENTS,
+        TYPE_PARAM_ON_VARIANT_CTOR,
```
Comment (eddyb) [inline src/librustc_session/lint/builtin.rs]: I'm not sure I would include `CTOR` in the name. Also, this isn't limited to *type* parameters.

I'd call it `VARIANT_GENERICS` or `VARIANT_GENERIC_ARGS`.

File: src/librustc_typeck/astconv.rs [hunk 6]
```diff
@@ -2305,13 +2313,36 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
...
+        self.tcx()
+            .sess
+            .struct_span_err(
+                segments_with_params,
+                "multiple segments with type parameters are not allowed",
+            )
+            .note("only a single path segment can specify type parameters")
+            .emit();
+        true
+    }
```
Comment (eddyb) [inline src/librustc_typeck/astconv.rs]: I think this is unnecessarily complicated, it doesn't distinguish this "`enum` vs variant" situation from others (e.g. trying to pass generic args to a `mod`) and it's wrong with generic associated items like methods or GATs.

This could be a help message on the error from `prohibit_generics`, such as pointing to the `enum` segment and saying "this segment *already* specified generic args for the `enum`" or something.

File: src/librustc_typeck/astconv.rs [hunk 7]
```diff
@@ -2504,18 +2534,51 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
...
+    fn lint_type_param_on_variant_ctor(&self, segments: &[hir::PathSegment<'_>]) {
+        // Doing this to get around rustfmt-caused line too long.
+        use hir::PathSegment as P;
+        if let [.., prev, P { ident, hir_id: Some(hir_id), args: Some(args), .. }] = segments {
+            let sp = args.span;
+            if sp.hi() == sp.lo() || sp == DUMMY_SP || sp.parent().is_some() {
+                // The params were not written by the user, but rather derived. These are expected.
+                // `Enum::Variant` where `Variant` has inferred lifetimes.
+                return;
+            }
```
Comment (eddyb) [inline src/librustc_typeck/astconv.rs]: I don't like this check, and you can get rid of the lifetimes for now (they shouldn't be added to value paths anyway).

File: src/librustc_typeck/astconv.rs (hunk 6 above)
Comment (estebank) [reply src/librustc_typeck/astconv.rs]: > I think this is unnecessarily complicated, it doesn't distinguish this "enum vs variant" situation from others (e.g. trying to pass generic args to a mod) and it's wrong with generic associated items like methods or GATs.

That check [is being done when called](https://github.com/rust-lang/rust/pull/69363/files#diff-1d1b0d29a2e8da97c6bfb6e364d920c7R5420). If it were called without that check it would indeed be incorrect.

> This could be a help message on the error from prohibit_generics, such as pointing to the enum segment and saying "this segment already specified generic args for the enum" or something.

Wanted to separate this from E0109 to potentially give it its own error code to explain that this will never be supported as opposed to the more generic current error.

File: src/librustc_typeck/astconv.rs (hunk 7 above)
Comment (estebank) [reply src/librustc_typeck/astconv.rs]: Agree. It's hacky as hell.

> you can get rid of the lifetimes for now

How?

> they shouldn't be added to value paths anyway

I've spent too much time trying to figure out of a way of making this change without affecting currently accepted code.

File: src/librustc_typeck/astconv.rs (hunk 6 above)
Comment (eddyb) [reply src/librustc_typeck/astconv.rs]: > to explain that this will never be supported

What do you mean? Generics on both the `enum` and a variant is a plausible feature (GADT), and generics on modules have been talked about in the past.

The error message looks pretty misleading to me, it seems to claim some absolute truth.

File: src/librustc_typeck/astconv.rs (hunk 7 above)
Comment (eddyb) [reply src/librustc_typeck/astconv.rs]: Basically you want all `Expr::Struct` with variant paths to behave in HRI lowering like they did when you had removed the `DefKind::Variant` arm.
You don't want the elision lifetimes being injected *at all*.

They're for `fn foo(x: &T) -> Enum::Variant`, or perhaps `<Enum>::Variant {...}` (neither of which compile right now), *not* `Enum::Variant {...}`.

Comment (estebank) [reply src/librustc_typeck/astconv.rs]: But if none of the PathSegments _have_ the elision lifetimes injected (in src/librustc_ast_lowering/path.rs) then you get knock down errors about the type for a HirId not existing .

Comment (eddyb) [reply src/librustc_typeck/astconv.rs]: > then you get knock down errors about the type for a HirId not existing

I don't recall you bringing this up, I thought it just worked while you had that arm removed? I don't see why it would, other value paths have no lifetimes injected into them *at all*.

**Comment by Dylan-DPC** [comment]
@estebank waiting on your answers to the latest questions on the above review. Thanks

**Comment by Dylan-DPC** [comment]
Closing this due to inactivity
